imap_processing/codice/codice_l1a.py

```diff
@@ -12,18 +11,299 @@
+        )
+        sweep_data = pd.read_csv(esa_sweep_data_file)
+
+        # Determine which ESA sweep table is needed
+        sweep_table_id = ESA_SWEEP_TABLE_ID_LOOKUP[(self.plan_id, self.plan_step)]
+
+        # Get the appropriate values
+        sweep_table = sweep_data[sweep_data["table_idx"] == sweep_table_id]
+        self.esa_sweep_values = sweep_table["esa_v"].values
+
+    def get_lo_data_products(self):
+        """Retrieve the lo data products table."""
+        # TODO: implement this
+
+        pass
+
+    def make_cdf_data(self):
+        """Create the ``xarray`` datasets needed for the L1a CDF file.
+
+        Returns
+        -------
+        dataset : xr.Dataset
+            The dataset used in the L1a CDF file
+        """
+        # TODO: Properly implement this
+
+        # metadata_arrays = collections.defaultdict(list)
+        #
+        epoch_time = xr.DataArray(
+            [np.datetime64("2010-01-01T00:01:06.184")],
+            name="Epoch",
+            dims=["Epoch"],
+            attrs=ConstantCoordinates.EPOCH,
+        )
+
+        # energy = xr.DataArray(self.esa_sweep_values, name="Energy", dims=["Energy"])
+        # times = xr.DataArray(self.acquisition_times, name="Times", dims=["Times"])
+
+        dataset = xr.Dataset(
+            coords={
+                "Epoch": epoch_time,
+                # "Energy": energy,
+                # "Times": times,
+            },
+            attrs=codice_l1a_global_attrs.output(),
+        )
+
+        # dataset["SCIENCE_DATA"] = self.science_values
+
+        return dataset
+
+    def unpack_science_data(self):
+        """Make 4D L1a data product from the decompressed science data.
+
+        Take the decompressed science data and reorganize the bytes to
+        create a four-dimensional data product
+
+        TODO: Describe the data product in more detail in docstring
+        """
+        print("Unpacking science data")
+
+        self.compression_algorithm = LO_COMPRESSION_ID_LOOKUP[self.view_id]
+        self.collapse_table_id = LO_COLLAPSE_TABLE_ID_LOOKUP[self.view_id]
+
+        # Generate simulated science data
+        # TODO: Take hard coded bit length value out and use variable instead
+        if self.use_simulated_data:
+            self.science_values = self._generate_simulated_data(37748736)
+        else:
+            # Decompress the science data
+            compressed_values = None
+            print(
+                f"Decompressing science data using {self.compression_algorithm.name} algorithm"  # noqa
+            )
+            self.science_values = [
+                decompress(compressed_value, self.compression_algorithm)
+                for compressed_value in compressed_values
+            ]
+
+        # TODO: Figure out how to properly unpack the data
+        # 128 e/q steps of 12 spin sectors x 5 positions
+        # Chunk the data by energy steps
+        num_bytes = len(self.science_values)
```

Comment:
is it number of bits? I don't remember now. I think there was mention of byte count.